ft_prepare_sourcemodel.m

```diff
@@ -166,96 +170,84 @@
-basedonshape      = ~isempty(cfg.headshape);                                                  % surface mesh based on inward shifted head surface from external file
-basedonmri        = isfield(cfg, 'mri') && ~(isfield(cfg, 'warpmni') && istrue(cfg.warpmni)); % regular 3D grid, based on segmented MRI, restricted to gray matter
-basedonmni        = isfield(cfg, 'mri') &&  (isfield(cfg, 'warpmni') && istrue(cfg.warpmni)); % regular 3D grid, based on warped MNI template
-basedonvol        = false;                                                                    % surface mesh based on inward shifted brain surface from volume conductor
-basedoncortex     = isfield(cfg, 'headshape') && (iscell(cfg.headshape) || any(ft_filetype(cfg.headshape, {'neuromag_fif', 'freesurfer_triangle_binary', 'caret_surf', 'gifti'}))); % cortical sheet from external software such as Caret or FreeSurfer, can also be two separate hemispheres
-basedonresolution = isfield(cfg, 'resolution') && ~basedonmri && ~basedonmni;                 % regular 3D grid with specification of the resolution
-basedonfile       = isfield(cfg, 'sourcemodel') && ischar(cfg.sourcemodel);
-
-if basedonshape && basedoncortex
-  % treating it as cortical sheet has preference
-  basedonshape = false;
-end
-
-if basedongrid && basedonpos
-  % fall back to default behavior, in which the pos overrides the grid
-  basedongrid = false;
-end
-
-if ~any([basedonresolution basedongrid basedonpos basedonshape basedonmri basedoncortex basedonmni]) && ~isempty(cfg.headmodel)
-  % fall back to default behavior, which is to create a surface grid (e.g. used in FT_MEGREALIGN)
-  basedonvol = true;
+if ~isfield(cfg, 'method') || isempty(cfg.method)
+  if isfield(cfg, 'xgrid') && ~ischar(cfg.xgrid)
+    cfg.method = 'basedongrid'; % regular 3D grid with explicit specification
+  elseif isfield(cfg.sourcemodel, 'pos')
+    cfg.method = 'basedonpos'; % using user-supplied positions, which can be regular or irregular
+  elseif ~isempty(cfg.headshape)
+    cfg.method = 'basedonshape'; % surface mesh based on inward shifted head surface from external file
+  elseif isfield(cfg, 'mri') && ~(isfield(cfg, 'warpmni') && istrue(cfg.warpmni))
+    cfg.method = 'basedonmri'; % regular 3D grid, based on segmented MRI, restricted to gray matter
+  elseif isfield(cfg, 'mri') &&  (isfield(cfg, 'warpmni') && istrue(cfg.warpmni))
+    cfg.method = 'basedonmni'; % regular 3D grid, based on warped MNI template
+  elseif isfield(cfg, 'headshape') && (iscell(cfg.headshape) || any(ft_filetype(cfg.headshape, {'neuromag_fif', 'freesurfer_triangle_binary', 'caret_surf', 'gifti'})))
+    cfg.method = 'basedoncortex'; % cortical sheet from external software such as Caret or FreeSurfer, can also be two separate hemispheres
+  elseif isfield(cfg, 'resolution') && ~basedonmri && ~basedonmni
```

Comment:
I might be mistaken, but the 'basedonsomething' arguments here don't exist, so why are they included in the elseif statements? Do they need to be replaced by someting more informative?

Reply:
You are correct, I missed this. I think these statements can be removed entirely, because the statement can't even be reached if cfg.method already is 'basedonmri', or 'basedonmni'
```suggestion
  elseif isfield(cfg, 'resolution')
```